### PR TITLE
Add codeql analysis workflow for JS

### PR DESCRIPTION
**Pull request by thelovekesh**
## Summary
Add CodeQL analysis for JS language.

Fixes task 6 of #973

## Relevant technical choices

- Add CodeQL official workflows.
- Add concurrency to cancel any in-progress workflow runs.
- Allow only `security:write` permissions in workflow.

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

### CODE REVIEW

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+on:
+  push:
+    # Only run if JS files changed.
+    paths:
+      - "**.js"
+    branches:
+      - trunk
+      - "release/**"
+  pull_request:
+    # Only run if JS or Python files changed.
```
Comment (westonruter) [inline .github/workflows/codeql-analysis.yml]:
```suggestion
    # Only run if JS files changed.
```

File: .github/workflows/codeql-analysis.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      - trunk
+      - "release/**"
+  pull_request:
+    # Only run if JS files changed.
+    paths:
+      - "**.js"
+    branches:
+      - trunk
+      - "release/**"
+      - "feature/**"
```
Comment (westonruter) [inline .github/workflows/codeql-analysis.yml]: I think this should apply to branches for all PRs. 

```suggestion
```

But it can be done in another PR along with the other workflows. See https://github.com/WordPress/performance/pull/972#discussion_r1487151134

**Review by westonruter — APPROVED** [review]

File: .github/workflows/codeql-analysis.yml (hunk 2 above)
Comment (thelovekesh) [reply .github/workflows/codeql-analysis.yml]: Fixed in https://github.com/WordPress/performance/pull/981/commits/f4e103c9fda1f5c92e9695a25910a333a3cdf859

**Comment by thelovekesh** [comment]
@joemcgill @felixarntz can you folks please take a look?

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -2,7 +2,7 @@
 /admin                                      @mukeshpanchal27 @felixarntz
 /tests/admin                                @mukeshpanchal27 @felixarntz
 /bin                                        @mukeshpanchal27 @felixarntz
-/.github                                    @mukeshpanchal27 @felixarntz
+/.github                                    @mukeshpanchal27 @felixarntz @thelovekesh
```
Comment (joemcgill) [inline .github/CODEOWNERS:5]: Not opposed, but am wondering if we should specify code owners specifically for workflows, rather than the whole directory?

File: .github/workflows/codeql-analysis.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+# Cancel previous workflow run groups that have not completed.
+concurrency:
+  # Group workflow runs by workflow name, along with the head branch ref of the pull request
+  # or otherwise the branch or tag ref.
+  group: ${{ github.workflow }}-${{ github.event_name == 'pull_request' && github.head_ref || github.ref }}
+  cancel-in-progress: true
+
+# Disable permissions for all available scopes.
+# Enable permissions for specific scopes as needed on job level.
+permissions: {}
```
Comment (joemcgill) [inline .github/workflows/codeql-analysis.yml:13]: I'm assuming these are copied from the props-bot workflow? Seems sensible to me, just noting that most of our workflows don't include concurrency groups or a removal of permissions scopes. I assume the concurrency group here is meant to cancel any in-progress jobs on PRs whenever a new commit is pushed to that branch or if the current workflow is restarted for some reason?

File: .github/workflows/codeql-analysis.yml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    # Only run if JS files changed.
+    paths:
+      - '**.js'
+
+jobs:
+  analyze:
+    name: Analyze
+    runs-on: ubuntu-latest
+    permissions:
+      security-events: write
```
Comment (joemcgill) [inline .github/workflows/codeql-analysis.yml:33]: Why is this needed?

**Review by joemcgill — COMMENTED** [review]
Thanks @thelovekesh. I've left a couple of comments, but don't see any blockers.

File: .github/CODEOWNERS (hunk 3 above)
Comment (thelovekesh) [reply .github/CODEOWNERS:5]: @joemcgill Not sure what would be more appropriate but I am okay to review any workflow file changes.

File: .github/workflows/codeql-analysis.yml (hunk 4 above)
Comment (thelovekesh) [reply .github/workflows/codeql-analysis.yml:13]: @joemcgill Most of the workflows rely on a third-party action to control the concurrency which is okay but I am not in favor of depending on third-party dependencies while something can be achieved natively, and yes it's here for cancelling the previous workflow run.

Also restricting the permission in the runner is a good practice to keep the metadata of a repo intact in case some action is compromised.

Comment (thelovekesh) [reply .github/workflows/codeql-analysis.yml:13]: Referenced from: https://github.com/ampproject/amp-wp/blob/develop/.github/workflows/build-test-measure.yml

File: .github/CODEOWNERS (hunk 3 above)
Comment (thelovekesh) [reply .github/CODEOWNERS:5]: I think this would be more appropriate as I will be working on optimizing most of the workflows under https://github.com/WordPress/performance/issues/973

File: .github/workflows/codeql-analysis.yml (hunk 5 above)
Comment (thelovekesh) [reply .github/workflows/codeql-analysis.yml:33]: Since CodeQL requires permission to write to security events as per https://github.com/github/codeql-action?tab=readme-ov-file#permissions

More info on how it's add to security events - https://github.com/WordPress/performance/security/code-scanning?query=pr%3A981+is%3Aopen

File: .github/workflows/codeql-analysis.yml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+# Disable permissions for all available scopes.
+# Enable permissions for specific scopes as needed on job level.
+permissions: {}
+
+on:
+  push:
+    # Only run if JS files changed.
+    paths:
+      - '**.js'
```
Comment (felixarntz) [inline .github/workflows/codeql-analysis.yml:19]: Does CodeQL only cover first-party JS? What about dependencies? Just curious, I'm not familiar with CodeQL.

I'm only asking since, if it also covers dependencies, we may also want to include `package.json` and `package-lock.json` here.

File: .github/workflows/codeql-analysis.yml (hunk 4 above)
Comment (felixarntz) [reply .github/workflows/codeql-analysis.yml:13]: @thelovekesh Your rationale makes sense to me, however I would be in favor of making such decisions consistently. I think for this PR we should stick with how we have been handling concurrency and permissions so far, as changing that is not the purpose of the PR. And then we can open a separate issue and/or PR to consider changing it holistically.

There is no good reason why this workflow should behave differently from the other ones in those points.

File: .github/workflows/codeql-analysis.yml (hunk 6 above)
Comment (thelovekesh) [reply .github/workflows/codeql-analysis.yml:19]: @felixarntz For dependencies, @dependabot creates security alerts that can be configured from -  https://github.com/WordPress/performance/security/dependabot

File: .github/workflows/codeql-analysis.yml (hunk 4 above)
Comment (thelovekesh) [reply .github/workflows/codeql-analysis.yml:13]: @felixarntz  TBH, I'm not sure if we should be prevented from writing optimized workflows by default just because existing workflows don't include those optimizations. In any case, as part of https://github.com/WordPress/performance/issues/973, I will be updating them.

**Review by swissspidy — APPROVED** [review]
